browser-extension/mv3/src/service-worker/services/requestProcessor.ts

```diff
@@ -0,0 +1,133 @@
+    if (
+      !IGNORED_HEADERS_ON_REDIRECT.some(
+        (header) => requestDetails.requestHeaders[header] || requestDetails.requestHeaders[header.toLowerCase()]
+      )
+    ) {
+      return;
+    }
+
+    const matchedRule = this.findMatchingRule(
+      [...this.cachedRules.redirectRules, ...this.cachedRules.replaceRules],
+      requestDetails.url
+    );
+
+    if (!matchedRule) {
+      return;
+    }
+
+    const ignoredHeaderValues = IGNORED_HEADERS_ON_REDIRECT.map((header) => ({
+      name: header,
+      value: requestDetails.requestHeaders[header] || requestDetails.requestHeaders[header.toLowerCase()],
+    }));
+
+    const redirectedUrl = this.getRedirectedUrl(requestDetails.url, matchedRule);
+
+    return this.updateRequestSpecificRules(tabId, requestDetails.url, {
+      action: {
+        requestHeaders: ignoredHeaderValues.map((header: { name: string; value: string }) => ({
+          header: header.name,
+          value: header.value,
+          operation: chrome.declarativeNetRequest.HeaderOperation.SET,
+        })),
+        type: chrome.declarativeNetRequest.RuleActionType.MODIFY_HEADERS,
+      },
+      condition: {
+        // Exact URL is not used because for replace rules, a marker query param is add which changes the URL so exact URL doesn't match
+        urlFilter: `|${redirectedUrl}`,
+        resourceTypes: [chrome.declarativeNetRequest.ResourceType.XMLHTTPREQUEST],
+        tabIds: [tabId],
+        requestMethods: [requestDetails.method.toLowerCase() as chrome.declarativeNetRequest.RequestMethod],
+        excludedInitiatorDomains: ["requestly.io", "requestly.com"],
+      },
+    });
+  };
+
+  private findMatchingRule = (rules: Rule[], url: string): Rule | undefined => {
+    return rules.find((rule) => rule.pairs.some((pair) => matchSourceUrl(pair.source, url)));
+  };
+
+  private getRedirectedUrl = (requestUrl: string, matchedRule: Rule) => {
+    const pair = matchedRule.pairs.find((pair) => matchSourceUrl(pair.source, requestUrl));
+    const redirectedUrl = pair.destination || requestUrl.replace(pair.from, pair.to);
+    if (redirectedUrl === requestUrl) {
+      return null;
+    }
+    return redirectedUrl;
+  };
+
+  private updateRequestSpecificRules = async (
+    tabId: number,
+    requestUrl: string,
+    ruleDetails: {
+      action: chrome.declarativeNetRequest.RuleAction;
+      condition: chrome.declarativeNetRequest.RuleCondition;
+    }
+  ) => {
+    let ruleId = parseInt(`${Date.now() % 1000000}${Math.floor(Math.random() * 1000)}`);
+
+    const sessionRulesMap = tabService.getData(tabId, TAB_SERVICE_DATA.SESSION_RULES_MAP);
+
+    let removeRuleIds = [];
+
+    if (sessionRulesMap?.[requestUrl]) {
+      ruleId = sessionRulesMap[requestUrl];
+      removeRuleIds.push(ruleId);
+    }
+
+    tabService.setData(tabId, TAB_SERVICE_DATA.SESSION_RULES_MAP, {
+      ...sessionRulesMap,
+      [requestUrl]: ruleId,
```

Comment:
Shouldn't ruleIds be an array or map of rulesIds because there might be case when initiator_domain and auth case applies on same URL

{
    [requestUrl]: {
        auth_header: ruleId,
        initiator_domain: ruleId,
    }
}